platform-api/che-core-api-auth/src/main/java/org/eclipse/che/security/oauth/RemoteOAuthTokenProvider.java

```diff
@@ -0,0 +1,65 @@
+/*******************************************************************************
+ * Copyright (c) 2012-2015 Codenvy, S.A.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   Codenvy, S.A. - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.che.security.oauth;
+
+
+import com.google.inject.name.Named;
+
+import org.eclipse.che.api.auth.oauth.OAuthTokenProvider;
+import org.eclipse.che.api.auth.shared.dto.OAuthToken;
+import org.eclipse.che.api.core.ConflictException;
+import org.eclipse.che.api.core.ForbiddenException;
+import org.eclipse.che.api.core.NotFoundException;
+import org.eclipse.che.api.core.ServerException;
+import org.eclipse.che.api.core.UnauthorizedException;
+import org.eclipse.che.api.core.rest.HttpJsonHelper;
+import org.eclipse.che.api.core.rest.shared.dto.Link;
+import org.eclipse.che.dto.server.DtoFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.inject.Inject;
+import javax.ws.rs.core.UriBuilder;
+import java.io.IOException;
+
+
+/**
+ * Allow get token from OAuth service over http.
+ */
+public class RemoteOAuthTokenProvider implements OAuthTokenProvider {
+    private static final Logger LOG = LoggerFactory.getLogger(RemoteOAuthTokenProvider.class);
+
+    @Inject
+    @Named("api.endpoint")
+    protected String apiEndpoint;
+
+    @Override
+    public OAuthToken getToken(String oauthProviderName, String userId) throws IOException {
+        if (!userId.isEmpty()) {
+                try {
+                    UriBuilder ub = UriBuilder.fromUri(apiEndpoint)
+                                              .path("/oauth/token/")
```

Comment:
Is it possible to use direct REST method reference here?
If it is you should use it! See [example](https://github.com/codenvy/che-core/blob/4.0/platform-api/che-core-api-machine/src/main/java/org/eclipse/che/api/machine/server/command/CommandService.java#L190-L195).
The only reason why it may not be used is when you have cycle dependency in your project.